darts/models/forecasting/tide_model.py

```diff
@@ -0,0 +1,526 @@
+        temporal_width
+            The width of the future covariate embedding space.
+        dropout
+            Dropout probability
+        **kwargs
+            all parameters required for :class:`darts.model.forecasting_models.PLForecastingModule` base class.
+
+        Inputs
+        ------
+        x
+            Tuple of Tensors `(x_past, x_future, x_static)` where `x_past` is the input/past chunk and
+            `x_future`is the output/future chunk. Input dimensions are `(batch_size, time_steps, components)`
+        Outputs
+        -------
+        y
+            Tensor of shape `(batch_size, output_chunk_length, output_dim, nr_params)`
+
+        """
+
+        super().__init__(**kwargs)
+
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        self.future_cov_dim = future_cov_dim
+        self.static_cov_dim = static_cov_dim
+        self.nr_params = nr_params
+        self.num_encoder_layers = num_encoder_layers
+        self.num_decoder_layers = num_decoder_layers
+        self.decoder_output_dim = decoder_output_dim
+        self.hidden_size = hidden_size
+        self.temporal_decoder_width = temporal_decoder_width
+        self.dropout = dropout
+        self.temporal_width = temporal_width
+
+        # residual block for input feature projection
+        # this is only needed when covariates are used
+        self.feature_projection = _ResidualBlock(
+            input_dim=future_cov_dim,
+            output_dim=temporal_width,
+            hidden_size=hidden_size,
+            dropout=dropout,
+        )
+
+        # original paper doesn't specify how to use past covariates
+        # we assume that they pass them raw to the encoder
+        historical_future_covariates_flat_dim = (
+            self.input_chunk_length + self.output_chunk_length
+        ) * (self.temporal_width if future_cov_dim > 0 else 0)
+        encoder_dim = (
+            self.input_chunk_length * (input_dim - future_cov_dim)
+            + historical_future_covariates_flat_dim
+            + static_cov_dim
+        )
+
+        self.encoders = nn.Sequential(
+            *[
+                _ResidualBlock(
+                    input_dim=encoder_dim,
+                    output_dim=encoder_dim,
+                    hidden_size=hidden_size,
+                    dropout=dropout,
+                )
+                for _ in range(num_encoder_layers)
+            ]
+        )
+
+        self.decoders = nn.Sequential(
+            *[
+                _ResidualBlock(
+                    input_dim=encoder_dim,
+                    output_dim=encoder_dim,
+                    hidden_size=hidden_size,
+                    dropout=dropout,
+                )
+                for _ in range(num_decoder_layers)
+            ],
+            # add decoder output layer
+            nn.Linear(
+                encoder_dim,
+                decoder_output_dim * self.output_chunk_length * self.nr_params,
+            ),
+        )
```

Comment:
I like this change. I also 100% agree that the paper is very cryptic. They even have an error in the architecture diagram, since you can't stack the covariate projection with the decoder output.